File: 24-swap-nodes-in-pairs/24-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None or head.next == None:return head
        dummy = ListNode(-1) ; dummy.next = head
        
        left = dummy ; first = dummy.next ; second = first.next ; right = second.next
        while True:
            first.next = right
            second.next = first
            left.next = second
            left = first
            first = right
            if first == None:
                break
            second = first.next
            if second == None:
                break
            right = second.next
        return dummy.next

File: 24-swap-nodes-in-pairs/24-swap-nodes-in-pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null){
            return head ;
        }
        ListNode ans  = new ListNode(-1) ;
        ListNode curr = ans ;
        ListNode p1 = head ;
        ListNode p2 = head.next ;
        while (p1 != null && p2 != null){
            curr.next = new ListNode(p2.val);
            if (p2.next != null){
                p2 = p2.next.next;
            } else{
                p2 = null;
            }
            curr = curr.next;
            curr.next = new ListNode(p1.val);
            if (p1.next != null){
                p1 = p1.next.next;
            } else {
                p1 = null;
            }
            curr = curr.next;
        }
        if (p1 == null){
            curr.next = p2;
        }
        else{
            curr.next = p1;
        }
        return ans.next;
    }
}